File: python/Question2/Question2.py
# You may change this function parameters
def findMaxProfit(numOfPredictedTimes, predictedSharePrices):
    # Participants code will be here
    maxi = 0  
    n = numOfPredictedTimes
    for k in range(2, n):
        profit = [[0 for i in range(k + 1)] 
                     for j in range(n)] 
          
        # Profit is zero for the first 
        # day and for zero transactions 
        for i in range(1, n): 
              
            for j in range(1, k + 1): 
                max_so_far = 0
                  
                for l in range(i): 
                    max_so_far = max(max_so_far, predictedSharePrices[i] -
                                predictedSharePrices[l] + profit[l][j - 1]) 
                                  
                profit[i][j] = max(profit[i - 1][j], max_so_far) 
          
        if profit[n - 1][k] >= maxi:
            maxi = profit[n - 1][k]
        
    return maxi

def main():
    line = input().split()
    numOfPredictedTimes = int(line[0])
    predictedSharePrices = list(map(int, line[1:]))

    answer = findMaxProfit(numOfPredictedTimes, predictedSharePrices)
    # Do not remove below line
    print(answer)
    # Do not print anything after this line

if __name__ == '__main__':
    main()

File: csharp/Question2/Question2.cs
using System.CodeDom.Compiler;
using System.Collections.Generic;
using System.Collections;
using System.ComponentModel;
using System.Diagnostics.CodeAnalysis;
using System.Globalization;
using System.IO;
using System.Linq;
using System.Reflection;
using System.Runtime.Serialization;
using System.Text.RegularExpressions;
using System.Text;
using System;

namespace Solution
{

    class Solution
    {

        // You may change this function parameters
        
        static int FindMaxProfit(int numOfPredictedTimes, int[] predictedSharePrices)
        {
            // Participant's code will go here

            int max = 0;
            for(int start = 2; start < numOfPredictedTimes; start++){
                int[, ] sum = new int[start + 1, numOfPredictedTimes + 1]; 
          
                for (int i = 0; i <= start; i++) 
                    sum[i, 0] = 0; 
          
                for (int j = 0; j <= numOfPredictedTimes; j++) 
                    sum[0, j] = 0; 
          
                for (int i = 1; i <= start; i++) { 
                    for (int j = 1; j < numOfPredictedTimes; j++) { 
                        int max_so_far = 0; 
                        for (int m = 0; m < j; m++) 
                        max_so_far = Math.Max(max_so_far, predictedSharePrices[j] - 
                                       predictedSharePrices[m] + sum[i - 1, m]); 
                        sum[i, j] = Math.Max(sum[i, j - 1], max_so_far); 
                    } 
                } 
                max = sum[start, numOfPredictedTimes - 1] > max ? sum[start, numOfPredictedTimes - 1] : max;
            }
            return max; 

        }

        static void Main(string[] args)
        {
            String[] firstLine = Console.ReadLine().Split(" ");

            int[] firstLineIntArr = Array.ConvertAll(firstLine, sTemp => Convert.ToInt32(sTemp));

            int numOfPredictedTimes = firstLineIntArr[0];
            int[] predictedSharePrices = firstLineIntArr.Skip(1).ToArray();

            int result = FindMaxProfit(numOfPredictedTimes, predictedSharePrices);

            // Please do not remove the below line.
            Console.WriteLine(result);
            // Do not print anything after this line
        }
    }
}